### PR TITLE
[Bugfix][V1] Restore hybrid GDN prefix-cache hits under MTP spec decoding

### PR DESCRIPTION
**Pull request by Y-aang**
## Purpose

On **Qwen3.5-122B-A10B** with MTP speculative decoding, a replay of a cached prompt never reaches the depth the prefix cache could give it, and prompts whose length is a multiple of the hash unit get **no hit at all**. Measured live before this change, with a 1072-token GDN page and `--prefix-match-unit 67`:

| prompt | cached tokens on replay | ceiling |
|---|---|---|
| 1072 | **0** | 938 |
| 2000 | **0** | 1876 |
| 2144 | **0** | 2010 |
| 3000 | 1072 | 2881 |
| 3500 | 2144 | 3417 |

### Root cause

Three rules compose badly. A lookup caps at `prompt_len - 1`, because the last token must be recomputed for logits. The MTP/EAGLE drafter then rewinds the hit one hash unit (67) further. But the producer publishes the GDN state at the prompt's own tail boundary, `prompt_len // 67 * 67`, which is at or above the cap — so the recurrent group cannot resume where the rewind lands, and the coordinator intersects the two groups down to the previous full page (or to zero).

The same cap breaks the full-attention side whenever `prompt_len` is a multiple of 67 — and every page multiple is, since `1072 = 16 × 67`. Its partial-tail entry is registered at `prompt_len // 67 * 67 == prompt_len`, which the `- 1` cap excludes, so full attention itself matches nothing below the page boundary and the hybrid hit collapses to 0. That is the `1072 → 0`, `2144 → 0` row above.

### Fix

Write-side only, all of it gated on MTP/EAGLE plus fine-grained hashing being active:

- `mamba_state_cache_position(num_tokens, block_size, hash_block_size)` in `kv_cache_utils.py`: `unit = min(hash_block_size, block_size)`, returns `(num_tokens - 1 - unit) // unit * unit` — the position a replay actually lands on.
- `Scheduler._mamba_block_aligned_split` stops a prefill chunk there, so that state exists to be published. It is an **extra** stop; the existing block-aligned stops stay. Replacing them lets intermediate chunks end off-grid and re-triggers the state poisoning of #43559 (14 tests in `test_mamba_align_chunk_split.py` fail that way).
- `MambaManager._cache_partial_tail_block` publishes the GDN state there instead of at the prompt tail.
- `FullAttentionManager._cache_partial_tail_block` additionally publishes `(prompt_len - 1) // 67 * 67`, the deepest tail a replay can reach. The unshifted tail is kept, because requests that *extend* this prompt have a higher cap and can still match it.
- `MambaManager` reports a block-aligned prompt end as unreachable through a new `reachable_block_mask()` argument, so that state is not cached in any retention mode: no lookup can land at or past a prompt's own end, and the rewound state below it is where consumers land. Guarded on the latter being cached. The rewound position is read from one `MambaManager` helper rather than recomputed per caller, and keys off whether the engine runs a drafter rather than the per-group `use_eagle` bit, which is only set on groups that take the drop.

Without the last item the page multiples stay at 0 no matter where the GDN state sits: the producer had the state at 938 / 2010 / 3082, but full attention offered nothing below 1072 / 2144 / 3216 for the intersection to meet.

## Test Plan

```bash
# unit
pytest tests/v1/core -q

# e2e: Qwen3.5-122B-A10B, 4x B200, MTP 3 tokens, sub-page hashing
vllm serve $QWEN35_122B --tensor-parallel-size 4 --enable-chunked-prefill \
    --spec-method mtp --spec-tokens 3 --prefix-match-unit 67 --port 8110
```

Hit depth is read from the `vllm:prefix_cache_hits_total` delta of a replay: a token-id prompt of an exact length is sent three times and the second and third deltas are the cached length. 132 lengths from 1 to 15k were swept — every page multiple with its `+1`, `+unit-1`, `+unit`, `+unit+1` neighbours, plus scattered lengths. Two further passes cover the configurations where the patch must do nothing, and quality was checked on top of real hits.

## Test Result

(1) Hit Length Boundary

All 132 lengths land exactly on the replay ceiling `(P - 1 - 67) // 67 * 67`. Page-boundary neighbourhoods, where the old code either missed entirely or fell back a full page:

| prompt | position | before | after | replay ceiling |
|---|---|---|---|---|
| 1072 | k*1072 | 0 | 938 | 938 |
| 1073 | k*1072 + 1 | 0 | 1005 | 1005 |
| 1138 | k*1072 + unit - 1 | 0 | 1005 | 1005 |
| 1139 | k*1072 + unit | 0 | 1005 | 1005 |
| 1140 | k*1072 + unit + 1 | 0 | 1072 | 1072 |
| 2144 | k*1072 | 0 | 2010 | 2010 |
| 2145 | k*1072 + 1 | 1072 | 2077 | 2077 |
| 2210 | k*1072 + unit - 1 | 1072 | 2077 | 2077 |
| 2211 | k*1072 + unit | 1072 | 2077 | 2077 |
| 2212 | k*1072 + unit + 1 | 1072 | 2144 | 2144 |
| 3216 | k*1072 | 0 | 3082 | 3082 |
| 3217 | k*1072 + 1 | 2144 | 3149 | 3149 |
| 3282 | k*1072 + unit - 1 | 2144 | 3149 | 3149 |
| 3283 | k*1072 + unit | 2144 | 3149 | 3149 |
| 3284 | k*1072 + unit + 1 | 2144 | 3216 | 3216 |
| 5360 | k*1072 | 0 | 5226 | 5226 |
| 5361 | k*1072 + 1 | 4288 | 5293 | 5293 |
| 5426 | k*1072 + unit - 1 | 4288 | 5293 | 5293 |
| 5427 | k*1072 + unit | 4288 | 5293 | 5293 |
| 5428 | k*1072 + unit + 1 | 4288 | 5360 | 5360 |
| 10720 | k*1072 | 7504 | 10586 | 10586 |
| 10721 | k*1072 + 1 | 9648 | 10653 | 10653 |
| 10786 | k*1072 + unit - 1 | 9648 | 10653 | 10653 |
| 10787 | k*1072 + unit | 9648 | 10653 | 10653 |
| 10788 | k*1072 + unit + 1 | 9648 | 10720 | 10720 |

Other lengths:

| prompt | before | after | ceiling |
|---|---|---|---|
| 500 | 0 | 402 | 402 |
| 999 | 0 | 871 | 871 |
| 2000 | 0 | 1876 | 1876 |
| 3000 | 1072 | 2881 | 2881 |
| 3500 | 2144 | 3417 | 3417 |
| 5000 | 3216 | 4891 | 4891 |
| 7777 | 6432 | 7705 | 7705 |
| 12345 | 10720 | 12261 | 12261 |
| 14999 | 12864 | 14874 | 14874 |

`before` comes from a CPU replica driving the real splitter and the real managers on the base commit; on the patched build that replica reproduces the end-to-end numbers length by length, so the two columns are comparable.

Configurations where the patch must be inert: re-running the same 132 lengths without `--prefix-match-unit` (both groups sit at 1072, so `hash_block_size == block_size` and sub-page hashing is off) gives `((P - 1) // 1072) * 1072 - 1072` on every length, and the CPU replica shows this build and the base commit agreeing on all 132. `mamba_cache_mode=all` cannot be exercised on this model — it is not in `SupportsMambaPrefixCaching`, so vLLM falls back to `align`.

Unit tests: `tests/v1/core` 492 passed. The one failure and two errors are GPU-exclusive e2e cases (`Free memory on device cuda:0 ... less than desired`) blocked by the serving process on the same box, and reproduce without this change. `tests/v1/core/test_mamba_align_chunk_split.py` is updated so the replay landing position counts as a legal intermediate chunk end alongside the partial-tail boundary.

(2) Model Quality Test

GPQA Diamond Score for Qwen/Qwen3.5-122B-A10B : 85.9% (match official score 86.6%)

GSM8K 5-shot, 300 questions: 0.890 / 0.853 with 0.0% / 1.3% invalid on this build, against 0.847 / 0.830 with 1.7% / 2.7% invalid on the base commit. Run-to-run spread is larger than the gap, so this is "no regression", not an accuracy claim.

Output quality on top of a real hit: an English passage of 700 / 1500 / 2400 / 3500 words, sent cold and then replayed (hits of 670 / 1608 / 2613 / 3819 tokens), returns byte-identical text at `temperature=0` in all four, twice over — no truncation, no repetition, no garbling.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

## Reproduce Low KVCache Hit Length with Qwen3.5

(1) Launch

```bash
CUDA_VISIBLE_DEVICES=0,1,2,3 \
PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True \
vllm serve Qwen/Qwen3.5-122B-A10B \
    --served-model-name Qwen3.5-122B-A10B \
    --trust-remote-code \
    --tensor-parallel-size 4 \
    --gpu-memory-utilization 0.90 \
    --max-model-len 262144 \
    --max-num-batched-tokens 8192 \
    --max-num-seqs 64 \
    --enable-chunked-prefill \
    --spec-method mtp \
    --spec-tokens 3 \
    --prefix-match-unit 67 \
    --host 0.0.0.0 --port 8110
```

(2) Send the same prompt twice

```bash
python3 - <<'EOF'
import json, random, re, urllib.request

N = 2144
BASE = "http://localhost:8110"

def hits_total():
    metrics = urllib.request.urlopen(f"{BASE}/metrics").read().decode()
    return float(
        re.search(r"^vllm:prefix_cache_hits_total\{[^}]*\} (\S+)", metrics, re.M).group(1)
    )

random.seed(N)
body = json.dumps({
    "model": "Qwen3.5-122B-A10B",
    "prompt": [random.randint(1000, 100000) for _ in range(N)],
    "max_tokens": 1,
    "temperature": 0,
}).encode()

for attempt in (1, 2):
    before = hits_total()
    urllib.request.urlopen(urllib.request.Request(
        f"{BASE}/v1/completions", body, {"Content-Type": "application/json"}))
    print(attempt, int(hits_total() - before))
EOF
```

(3) Check the second request's cache length

Before the fix it is 0, after the fix it is 2010.

### CODE REVIEW

**Comment by ZJY0516** [comment]
Thanks for your contribution.

My concern is it will introduce an extra chunk and forward, which will hurt TTFT performance

For example,   `block_size = 8, prefix unit = 4, prompt_len = 18`

```
  main：

  0 → 16 → 18

  PR：

  0 → 12 → 16 → 18
```

**Comment by ZJY0516** [comment]
https://github.com/vllm-project/vllm/pull/50897 is a more proper fix, but it's not easy to land it

**Comment by sungsooha** [comment]
We validated this PR on **NVIDIA-Nemotron-3-Ultra-550B-A55B-NVFP4** (hybrid Mamba2 + full
attention, MTP-5 speculative decoding, disaggregated prefill/decode on 8×B200, vLLM 0.27.2rc1 +
Dynamo). Sharing measurements since the PR currently has no serving numbers.

**1. `mamba_state_cache_position` is exact.** We confirmed
`cached = max((n−1−U)//U × U, 0)` on **nine independent configurations** — three prompt lengths ×
three prefix-match units (4416 / 1104 / 276) — matching the observed
`usage.prompt_tokens_details.cached_tokens` **to the token** every time.

**2. On a config with a single KV block size, the patch is inert without `--prefix-match-unit`.**
Our two KV groups both resolve to block 4416, so `hash_block_size = gcd = 4416 = block_size`,
`_cache_partial_tail_block` returns early, and the scheduler `state_position` stop is gated off via
`mamba_partial_cache_hit`. Worth stating in the PR description — we initially expected a standalone
effect and measured none. (`_unreachable_boundaries` can still fire for exact-block-multiple
prompts.)

**3. Paired serving result** (`--prefix-match-unit 1104`, i.e. 16 units per 4416-token block —
the same block:unit ratio this PR validated at 1072/67), vs an otherwise identical control, 4-rung
concurrency ladder, cold cache per rung:

| cc | prefix-hit | tok/s/GPU | TTFT avg |
|---|---|---|---|
| 16 | +9.09 pp | +3.8% | −13.6% |
| 32 | +9.12 pp | +7.4% | −30.7% |
| 64 | +9.17 pp | +15.3% | −37.4% |
| 128 | +6.99 pp | +19.9% | −45.6% |

Strictly better on **both** throughput and TTFT at every rung. At matched TTFT the frontier moves
**+20% to +46%**. Correctness: 6/6 byte-identical cold-vs-warm greedy outputs with
`cached_tokens > 0` on every warm request.

**4. One cost worth documenting — the scheduler stop under-packs prefill batches.** The mandatory
stop at `state_position` truncates the chunk, so iterations carry **56%** of the
`max_num_batched_tokens` budget versus **84%** on the control. Iterations per request rise ~14%
while **per-iteration wall time is unchanged**, so prefill wall time is governed by iteration count
rather than token count: prefill throughput falls **−26%** at cc=16 to **−47%** at cc=128, even
though 22–25% *fewer* tokens are computed.

This is out-earned by the queueing win in our workload, but it is not second-order. We separated it
from hash-chain cost by running U=276 vs U=1104 (4× fewer hash entries): that recovered only 3.6 of
31.3 points, while both units showed the *same* ~14% iteration inflation — so it is a per-request
step cost from the extra stop, not a per-hash-entry cost.

**5. Minor:** `_cache_partial_tail_block` builds `boundaries` as `tuple[int]` and then widens it
with `+= (replay_tail,)`; mypy rejects this. `boundaries: tuple[int, ...]` fixes it.

**Comment by jschmied** [comment]
Comment retracted: measurement was invalid, i will measure again

**Comment by jschmied** [comment]
Correcting my retracted comment. The harness was broken (a `cp -a` venv keeps absolute shebangs, so
every arm ran unpatched code); re-measured properly, with a log line inside
`_mamba_block_aligned_split` proving which tree ran.

The defect confirmation stands: on GB10, MTP n=3, block 1600, a prompt of exactly 6 400 tokens
caches **3 200** where the back-off alone predicts 4 800 — your `1072 → 0` row on different
hardware.

The fix still does not move it here: 3 200 / 4 800 on both merge-base and branch. The
instrumentation says why — `last_cache_position` stays 4 800 in both, so the back-off is still what
caps this case, and that is not what this PR changes. For contrast, #50897 on the same rig moves
the same two prompts to 4 800 / 6 400.

_Disclosure: AI-assisted analysis (Claude Code); I reviewed the arithmetic and code paths myself._

**Comment by Windless84** [comment]
Independent confirmation of the mechanism on Qwen3.8-Flash-Next NVFP4 (GDN + QSA + MTP-3, one RTX PRO 6000, 1,600-token pages, no `--prefix-match-unit`, so hash unit = page), measured by driving the real scheduler and managers over the model's actual 76-layer spec list on CPU and then live:

- With the trailing-block drop on, the first request extending a cached prompt hits **0**: the attention groups offer the last full page minus one, while the align-mode Mamba manager only holds its checkpoint on the last full page whenever the prompt's tail spans more than one prefill chunk (60,000 -> 80,000 tokens: 0). Only the *second* extension hits, through the reported uncached common prefix (80,000 -> 100,000: 78,400).
- With `disable_eagle_block_drop` the two line up and the first extension hits (60,000 -> 80,000: 59,200; 80,000 -> 100,000: 78,400). Live replay on that setting: 41K after 34K served 33,600 cached tokens (0.88 s vs 4.0 s), 250K after 140K served 139,200 (11.5 s vs 25.6 s). MTP acceptance on 400-token prose was unchanged by reusing the trailing block (47-52% either way), so we are running with the drop off as a stand-in until a write-side fix like this one lands. +1 for the rebase.

The drafter-group identification needed for the Mamba groups to be lookup-eligible at all on this model is in #56026.

**Comment by rufftruffles** [comment]
Supporting live data from #54094 with #56026 applied first. I did not apply the current #52244 head, so this is mechanism evidence rather than a current-head patch validation.

On Qwen3.8-27B + DFlash2 K=7 with sparse retention `0`, #56026 removed the unannotated-drafter warning and left Mamba outside the draft groups. A byte-identical 50,011-token replay still got zero hits with the default trailing-block drop:

```text
queries_total=100022
hits_total=0
cold TTFT=11.456 s
replay TTFT=10.435 s
```

Changing only `disable_eagle_block_drop=true` recovered the cache immediately:

```text
queries_total=100022
hits_total=46144
cold TTFT=10.543 s
replay TTFT=1.200 s
```

All four requests returned 6/6 buried facts. This isolates the remaining miss to the trailing-block-drop write side and supports the mechanism addressed by #52244. The PR currently reports merge conflicts, so I am not claiming its present head was tested.
